chain/chain/src/state_sync.rs

```diff
@@ -0,0 +1,191 @@
+use near_primitives::types::EpochId;
+use near_store::{DBCol, Store, StoreUpdate};
+
+use borsh::BorshDeserialize;
+use std::collections::BinaryHeap;
+
+use crate::types::BlockHeader;
+use crate::ChainStoreAccess;
+
+// Used for storing block headers in a heap that returns the lowest heights first
+#[derive(PartialEq, Eq)]
+struct BlockHeaderByHeight<'a>(&'a BlockHeader);
+
+impl<'a> PartialOrd for BlockHeaderByHeight<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<'a> Ord for BlockHeaderByHeight<'a> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other.0.height().cmp(&self.0.height())
+    }
+}
+
+fn get_epoch_new_chunks(
+    store: &Store,
+    block_hash: &CryptoHash,
+) -> Result<Option<Vec<usize>>, Error> {
+    Ok(store.get_ser(DBCol::StateSyncNewChunks, block_hash.as_ref())?)
+}
+
+fn iter_epoch_new_chunks_keys<'a>(
+    store: &'a Store,
+) -> impl Iterator<Item = Result<CryptoHash, std::io::Error>> + 'a {
+    store
+        .iter(DBCol::StateSyncNewChunks)
+        .map(|item| item.and_then(|(k, _v)| CryptoHash::try_from_slice(&k)))
+}
+
+fn iter_state_sync_hashes_keys<'a>(
+    store: &'a Store,
+) -> impl Iterator<Item = Result<EpochId, std::io::Error>> + 'a {
+    store
+        .iter(DBCol::StateSyncHashes)
+        .map(|item| item.and_then(|(k, _v)| EpochId::try_from_slice(&k)))
+}
+
+fn save_epoch_new_chunks<T: ChainStoreAccess>(
+    chain_store: &T,
+    store_update: &mut StoreUpdate,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let Some(mut num_new_chunks) = get_epoch_new_chunks(chain_store.store(), header.prev_hash())?
+    else {
+        // This might happen in the case of epoch sync where we save individual headers without having all
+        // headers that belong to the epoch.
+        return Ok(());
+    };
+
+    // This shouldn't happen because block headers in the same epoch should have chunks masks
+    // of the same length, but we log it here in case it happens for some reason. We return Ok because if this
+    // happens, it's some bug in this state sync logic, because the chunk mask length of headers are checked when
+    // they're verified. So in this case we shouldn't fail to commit this store update and store this block header
+    if num_new_chunks.len() != header.chunk_mask().len() {
+        tracing::error!(
+            block_hash=%header.hash(), chunk_mask_len=%header.chunk_mask().len(), stored_len=%num_new_chunks.len(),
+            "block header's chunk mask not of the same length as stored value in DBCol::StateSyncNewChunks",
+        );
+        return Ok(());
+    }
+
+    let mut done = true;
+    for (shard_index, num_new_chunks) in num_new_chunks.iter_mut().enumerate() {
+        let new_chunk = header.chunk_mask()[shard_index];
+        if new_chunk {
+            *num_new_chunks += 1;
+        }
+        if *num_new_chunks < 2 {
+            done = false;
+        }
```

Comment:
It's a usize but yeah might as well, agreed

Reply:
actually wait now that you mention it, let's change it to u8 instead of usize. I think usize will probably be fine since the size of the type is not going to change within the same system, but just feels like a weird thing to do to store a usize rather than one w an explicit size